### PR TITLE
[RNMobile] FloatingToolbar

### PR DESCRIPTION
**Pull request by lukewalczak**
## Description
PR is adding `FloatingToolbar` component without inner functionality.
gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1362
## How has this been tested?

Tested on both platforms

## Screenshots 

ios | android
--- | ---
 | 

## Types of changes

Added `FloatingToolbar`.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/block-editor/src/components/block-list/block.native.js
Comment (pinarol) [inline packages/block-editor/src/components/block-list/block.native.js]: We don’t have to exclude Media & Text block. We want the same UX everywhere InnerBlocks are used

File: packages/block-editor/src/components/block-list/block.native.js
Comment (pinarol) [inline packages/block-editor/src/components/block-list/block.native.js]: It should be enough to just check for inner blocks, no need to check if it is a group block

File: packages/block-editor/src/components/block-list/block.native.js
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/block.native.js]: Makes sense. Do we want to display toolbar above each `Media & Text` children when `Media & Text` is nested in `Group` component or only over the whole `Media & Text` component?

File: packages/block-editor/src/components/block-list/block.native.js
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/block.native.js]: Agree - `const hasRootInnerBlocks = rootBlock.innerBlocks.length !== 0;` should be enough.

File: packages/block-editor/src/components/block-list/block.native.js
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/block.native.js]: I've attached a gif presenting how `FloatingToolbar` behaves once I remove Media&Text exclusion
  

Please let me know if it's correct.

Comment (pinarol) [reply packages/block-editor/src/components/block-list/block.native.js]: Right, we need to show it whenever a block has InnerBlocks. If you check out the prototype link in the issue same thing can be seen with the column block.

**Comment by lukewalczak** [comment]
`FloatingToolbar` for the first item in a group is located outside the `FlatList` which is causing an issue that toolbar is not clickable - actually clicking the toolbar is focusing the previous block.

**Comment by lukewalczak** [comment]
Finally, we've managed to make it work as expected.

**Review by pinarol — COMMENTED** [review]
I went over the changes quickly but didn't have chance to test it yet so I'll re-review one more time.

Thanks a lot for the workaround for that issue about Floating Toolbar not receiving the events. 🙇  Could you please also create a new branch with the problematic scenario to later use for reporting a bug to react native repo?

**Comment by lukewalczak** [comment]
I'm still working on it since workaround doesn't work on Android and it still has the same issue :(

**Comment by mkevins** [comment]
Hi @lukewalczak :wave: 

I've been investigating the issue on Android as well, and would like to note some things I've found, in case it is useful. I noticed that when the outer block is selected, the overflowing floating toolbar is tappable, and working as expected:



I debugged the code to confirm that in this case, the `FloatingToolbar.Slot` is added in the `BlockListBlock` component, with the adjacent block (the block that the `FloatingToolbar` overlaps) as a direct sibling to the slot. This seems to be the same way with other _inner_ blocks, when their `FloatingToolbar` overlaps their neighbor.

**Comment by mkevins** [comment]
Another thing we may need to keep in mind while considering the workaround using the `isFirstBlock` flag is that this may not always be sufficient in all cases. Particularly, this could pose a problem when inner blocks are side-by-side (i.e. outer-block `flex-direction: row`). As an example, here is what the media-text block looks like with this approach:



I am not sure why the `FloatingToolbar.Slot` is not working as expected when rendered prior to `KeyboardAwareFlatList` in `BlockList`. Some things I tried, which did not work:

* Set `removeClippedSubviews` on the outer view of `BlockList`
* Preprend the `View` in `BlockList` with the `FloatingToolbar.Slot` and wrap them with a fragment
* I even tried hard-coding the `FloatingToolbar` & `Slot` prepending the `InnerBlocks` component in `GroupEdit` just to see what would happen: taps were still falling through to the view underneath

**Comment by koke** [comment]
I was trying to understand the issues better, and while I don't have any good suggestions yet, I noticed this while using the UI inspector in Xcode, which doesn't look right:

Screen Shot 2019-09-30 at 15 58 15

**Comment by lukewalczak** [comment]
@koke Can it be caused by a high `zIndex` value (100) ?

File: packages/block-editor/src/components/block-list/block.native.js
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/block.native.js]: Sure, `showFloatingToolbar` has more sense and keep the convention.

File: packages/block-editor/src/components/block-list/block.native.js
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/block.native.js]: It's a leftover after the previous solutions.

File: packages/block-editor/src/components/block-list/block.native.js
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/block.native.js]: `toolbarHeight` won't be needed anymore.

File: packages/block-editor/src/components/block-list/block-mobile-floating-toolbar.native.scss
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/block-mobile-floating-toolbar.native.scss]: Sure!

**Comment by koke** [comment]
> @koke Can it be caused by a high `zIndex` value (100) ?

No, it seems `FloatingToolbar` is being rendered for every block:

Screen Shot 2019-10-01 at 11 34 47

**Comment by lukewalczak** [comment]
@koke Thanks! The problem was that `FloatingToolbar` was rendered actually for every block, but now I've made it conditional to flag `showDisplayToolbar`.

File: packages/block-editor/src/components/block-list/index.native.js
Comment (pinarol) [inline packages/block-editor/src/components/block-list/index.native.js]: Do we still need this considering the current temporary fix? If we don't let's remove it to avoid having unexpected side effects

File: packages/block-editor/src/components/block-list/index.native.js
Comment (pinarol) [inline packages/block-editor/src/components/block-list/index.native.js]: Do we still need to change this file? considering we are rendering the first one inside the bounds. I know this is a temp fix but still I'd want to change minimum code to avoid any side effects until we find a better solution.

**Review by pinarol — COMMENTED** [review]
Thanks for the effort @lukewalczak Just dropping a few comments, aside from that I think it looks good.

About the re-occuring FloatingToobar problem @koke reported, could you open an issue? I don't think we should solve it with this PR but we should track it, there might be something wrong with slot-fill. Considering that we might want to change the FloatingToolbar designs let's not block this PR with that issue.

File: packages/block-editor/src/components/block-list/index.native.js
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/index.native.js]: Yes, it's still needed to get working properly `FloatingToolbar` on Android.

File: packages/block-editor/src/components/block-list/index.native.js
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/index.native.js]: If we want to simplify it to the minimum we don't need using slot/fill, however, I've decided to keep this for the next investigations.

**Comment by mkevins** [comment]
The temporary workaround seems to keep the floating toolbar within the bounds of the parent, so taps are not "falling through" anymore. But, the toolbar currently prevents visibility of the underlying content, which is problematic:



Is this expected behavior for this temporary iteration, or are we expecting the content to clear the toolbar?

cc: @pinarol

**Comment by pinarol** [comment]
> Is this expected behavior for this temporary iteration

Yes, this won’t go to prod as is, we’ll have another iteration. For now we are trying to unblock dependent PRs.

**Comment by pinarol** [comment]
@lukewalczak Could you open an issue with the screenshot here just for us to not forget about it: https://github.com/WordPress/gutenberg/pull/17399#issuecomment-536617259 ? After that it seems ready to merge to me.

**Review by mkevins — APPROVED** [review]
Approving for first iteration, so other dependent PRs can move forward. :+1:

**Comment by pinarol** [comment]
> @lukewalczak Could you open an issue in the gutenberg-mobile repo with the screenshot here just for us to not forget about it: #17399 (comment) ? After that it seems ready to merge to me.

Taking this back because I tried with the latest version and it is all fixed

**Review by pinarol — APPROVED** [review]
LGTM 🎉 We'll do separate iterations to make this better, for now let's merge for not blocking other PRs

**Pull request merged**
